### PR TITLE
Add experimental option to reduce extremities.

### PR DESCRIPTION
**Pull request by erikjohnston**
Adds new config option `cleanup_extremities_with_dummy_events` which
periodically sends dummy events to rooms with more than 10 extremities.

THIS IS REALLY EXPERIMENTAL.

**Note**: This currently will send the event to clients ~~and other servers~~.

### CODE REVIEW

File: synapse/events/__init__.py [hunk 1]
```diff
@@ -92,6 +92,18 @@ def is_soft_failed(self):
         """
         return getattr(self, "soft_failed", False)
 
+    def should_proactively_send(self):
+        """Whether the eventm, if ours, should be sent to other clients and
```
Comment (richvdh) [inline synapse/events/__init__.py]: eventm

File: synapse/handlers/message.py [hunk 2]
```diff
@@ -261,6 +261,16 @@ def __init__(self, hs):
         if self._block_events_without_consent_error:
             self._consent_uri_builder = ConsentURIBuilder(self.config)
 
+        if (
+            not self.config.worker_app
+            and self.config.cleanup_extremities_with_dummy_events
+        ):
+            # XXX: Send dummy events.
```
Comment (richvdh) [inline synapse/handlers/message.py]: XXX?

File: synapse/handlers/message.py [hunk 3]
```diff
@@ -261,6 +261,16 @@ def __init__(self, hs):
...
+            self.clock.looping_call(
+                self._send_dummy_events_to_fill_extremities,
```
Comment (richvdh) [inline synapse/handlers/message.py]: can you wrap this with `run_as_background_process` please

File: tests/storage/test_cleanup_extrems.py [hunk 4]
```diff
@@ -222,3 +222,44 @@ def test_forked_graph_cleanup(self):
...
+            self.create_and_send_event(
+                self.room_id, self.user, prev_event_ids=[event_id_start]
+            )
+
+        latest_event_ids = self.get_success(
+            self.store.get_latest_event_ids_in_room(self.room_id)
+        )
+        self.assertEqual(len(latest_event_ids), 50)
+
+        # Bump the reacto repeatedly so that the background updates have a
```
Comment (richvdh) [inline tests/storage/test_cleanup_extrems.py]: reacto!

**Comment by ara4n** [comment]
@erikjohnston so what is the answer to:

> but it's going to explode horribly when the user you pick turns out not to have any rights to send events in the room.

Would we need to special case auth rules for dummy events?

**Comment by ara4n** [comment]
Conclusion from IRLing with Erik:

> but it's going to explode horribly when the user you pick turns out not to have any rights to send events in the room.

 * it shouldn't explode
 * it just won't cleanup extremities in rooms where you can't speak.
 * unfortunately, rooms where you can't speak are ones where extremities may well accumulate the most
 * changing auth rules to allow anyone to send m.dummy events could cause more DoS problems than it solves
 * we might be better off combining both the dummy event solution as well as the "prune useless extremities" solution (https://github.com/matrix-org/synapse/issues/5438) in the longer term to catch that scenario.

**Pull request merged**